### PR TITLE
Lesson07

### PR DESCRIPTION
**Pull request by iliamikheenko**

### CODE REVIEW

File: lesson07/src/StatisticApp.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+public class StatisticApp {
+    public static void main(String[] args) {
+        String bookTitle;
+
+        System.out.print("Enter book title: ");
+
+        try (Scanner scanner = new Scanner(System.in)){
+             bookTitle = scanner.nextLine();
+        } catch (Exception e) {
```
Comment (shaitanaleksander) [inline lesson07/src/StatisticApp.java:14]: no need to use Scanner with try{} just for closing the scanner. Close the scanner at the end

File: lesson07/src/StatisticApp.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+        File book = new File("lesson07/src/book/" + bookTitle + ".txt");
+
+        if (!book.exists()){
+            System.out.println("Book doesn't exist in lesson07/src/book!");
+            return;
+        }
+
+        StatisticReader stReader = new StatisticReader(book);
+
+        String output =
```
Comment (shaitanaleksander) [inline lesson07/src/StatisticApp.java:27]: bad name ->  output not descriptive, better go with statisticTemplate or something like that

File: lesson07/src/StatisticReader.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+public class StatisticReader {
+    private final Map<String,Integer> wordCounter = new HashMap<>();
+    public StatisticReader(File book) {
+        this.textToWordCounter(book);
+    }
+    private void textToWordCounter(File book){
+        try (BufferedReader bookBR = new BufferedReader(new FileReader(book))){
+            String line;
+            Pattern pattern = Pattern.compile("[^a-zA-Z']");
```
Comment (shaitanaleksander) [inline lesson07/src/StatisticReader.java:21]: can be class level var

File: lesson07/src/StatisticReader.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+import java.util.Map;
+import java.util.regex.Pattern;
+
+public class StatisticReader {
+    private final Map<String,Integer> wordCounter = new HashMap<>();
+    public StatisticReader(File book) {
+        this.textToWordCounter(book);
+    }
+    private void textToWordCounter(File book){
+        try (BufferedReader bookBR = new BufferedReader(new FileReader(book))){
```
Comment (shaitanaleksander) [inline lesson07/src/StatisticReader.java:19]: bad var name

File: lesson07/src/StatisticReader.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+            Pattern pattern = Pattern.compile("[^a-zA-Z']");
+            while ((line = bookBR.readLine()) != null){
+                String[] words = pattern.split(line.trim().replaceAll("\\s+"," "));
+                for (String word : words) {
+                    word = word.toLowerCase();
+                    wordCounter.put(word,wordCounter.getOrDefault(word,0)+1);
+                }
+            }
+        } catch (IOException e) {
+            throw new RuntimeException(e);
```
Comment (shaitanaleksander) [inline lesson07/src/StatisticReader.java:30]: what is the point of catching the IOException and throwing RuntimeException?

File: lesson07/src/StatisticReader.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    public int countUniqueWords(){
+        return wordCounter.keySet().size();
+    }
+
+    public int countWords(){
+        return wordCounter.values().stream()
+                .mapToInt(s -> s)
+                .sum();
```
Comment (shaitanaleksander) [inline lesson07/src/StatisticReader.java:41]: better to use reduce. example Stream.of(1, 2, 4, 5, 6, 7, 8, 8, 9, 9).reduce(Integer::sum).orElse(0)

File: lesson07/src/StatisticWriter.java [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+import java.io.IOException;
+
+public class StatisticWriter {
+    public static void writeToFile(String text, String bookTitle) {
+        try ( BufferedWriter outputFile = new BufferedWriter(
+                new FileWriter("lesson07/src/book/" + bookTitle + "_statistic.txt"))){
+            outputFile.write(text);
+            outputFile.flush();
+        } catch (IOException e) {
+            throw new RuntimeException(e);
```
Comment (shaitanaleksander) [inline lesson07/src/StatisticWriter.java:14]: same here

**Review by shaitanaleksander — APPROVED** [review]

File: lesson07/src/StatisticApp.java (hunk 1 above)
Comment (iliamikheenko) [reply lesson07/src/StatisticApp.java]: Edited.

File: lesson07/src/StatisticApp.java (hunk 2 above)
Comment (iliamikheenko) [reply lesson07/src/StatisticApp.java]: Edited.
Renamed to statisticTemplate.

File: lesson07/src/StatisticReader.java (hunk 3 above)
Comment (iliamikheenko) [reply lesson07/src/StatisticReader.java]: Edited.

File: lesson07/src/StatisticReader.java (hunk 6 above)
Comment (iliamikheenko) [reply lesson07/src/StatisticReader.java]: Edited.

File: lesson07/src/StatisticWriter.java (hunk 7 above)
Comment (iliamikheenko) [reply lesson07/src/StatisticWriter.java]: Edited.
User will this massege:
"Failed writing a new file with statistic. Try again."

**Review by iliamikheenko — COMMENTED** [review]
checked

**Pull request merged**